### PR TITLE
[ir] Replace FuncCallExpression with FrontendFuncCallStmt

### PR DESCRIPTION
**Pull request by lin-hitonami**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7059
* __->__ #7027

issue: #6983
FuncCall is not just an expression. It executes code and returns the result. It should be more like FrontendAllocaStmt rather than being FuncCallExpression. If it is an expression, the IR printer handles it like an expression and prints the detail of the func call everywhere its return value is used. If it is a statement, it will only be printed once, and other statements/expressions reference its return value by its statement number.

### CODE REVIEW

File: taichi/ir/frontend_ir.cpp [hunk 1]
```diff
@@ -1332,6 +1310,14 @@ Expr ASTBuilder::expr_alloca() {
   return var;
 }
 
+Expr ASTBuilder::insert_func_call(Function *func, const ExprGroup &args) {
+  auto var = Expr(std::make_shared<IdExpression>(get_next_id()));
```
Comment (bobcao3) [inline taichi/ir/frontend_ir.cpp:1314]: Are functions guaranteed to return a value?

File: taichi/transforms/lower_ast.cpp [hunk 2]
```diff
@@ -82,6 +82,21 @@ class LowerAST : public IRVisitor {
     }
   }
 
+  void visit(FrontendFuncCallStmt *stmt) override {
+    auto block = stmt->parent;
+    auto ident = stmt->ident;
+    TI_ASSERT(block->local_var_to_stmt.find(ident) ==
+              block->local_var_to_stmt.end());
+    std::vector<Stmt *> args;
```
Comment (bobcao3) [inline taichi/transforms/lower_ast.cpp:90]:
```suggestion
    std::vector<Stmt *> args(stmt->args.exprs.size());
```

and switch the push back to indexing

File: taichi/transforms/lower_ast.cpp [hunk 3]
```diff
@@ -82,6 +82,21 @@ class LowerAST : public IRVisitor {
     }
   }
 
+  void visit(FrontendFuncCallStmt *stmt) override {
+    auto block = stmt->parent;
```
Comment (bobcao3) [inline taichi/transforms/lower_ast.cpp:86]: nit: Use the types here. `auto` doesn't deduce references and if the underlying type is a simple pointer it's easier to read & less likely to mess up

File: taichi/ir/frontend_ir.cpp (hunk 1 above)
Comment (lin-hitonami) [reply taichi/ir/frontend_ir.cpp:1314]: If it has no return value, it returns an empty struct.

Comment (bobcao3) [reply taichi/ir/frontend_ir.cpp:1314]: This feels iffy... Some IR might not like the idea of an empty struct

Comment (lin-hitonami) [reply taichi/ir/frontend_ir.cpp:1314]: Maybe we can deal with this in the codegen. We can return void in the codegen if there is no return value (the return type is an empty struct in CHI-IR). In this case the return value is not used anywhere so we don't need to care about its value.

Comment (bobcao3) [reply taichi/ir/frontend_ir.cpp:1314]: Maybe it's better to return a std::optional

**Review by bobcao3 — APPROVED** [review]
LGTM

**Pull request merged**